### PR TITLE
Test the role field of Trades

### PR DESCRIPTION
**Pull request by fa2a5qj3**
Ref: Missing trade role from getTrades() API https://github.com/haveno-dex/haveno/issues/1146
Ref: Apply trade role to getTrades result https://github.com/haveno-dex/haveno/pull/1147

Test with/without fix:

```
Jul-15 23:11:59:36 (+3489 ms)	[L0] BEFORE TEST "Do all trades contain role property"
Jul-15 23:11:59.559 [grpc-default-executor-0] INFO  h.d.g.i.CallRateMeteringInterceptor: GetTrades has been called 1 time in the last second, rate limit is 10/second 
Jul-15 23:11:59.684 [grpc-default-executor-0] INFO  h.d.g.i.CallRateMeteringInterceptor: GetTrades has been called 2 times in the last second, rate limit is 10/second 
Jul-15 23:11:59.753 [grpc-default-executor-0] INFO  h.d.g.i.CallRateMeteringInterceptor: GetTrades has been called 1 time in the last second, rate limit is 10/second 
Jul-15 23:11:59:35 (+299 ms)	[L0] Trades size: 6
Jul-15 23:11:59:35 (+0 ms)	[L0] Trade: 62002 role: XMR buyer as maker
Jul-15 23:11:59:36 (+1 ms)	[L0] Trade: 87102 role: XMR buyer as maker
Jul-15 23:11:59:36 (+0 ms)	[L0] Trade: 74839807 role: XMR buyer as maker
Jul-15 23:11:59:36 (+0 ms)	[L0] Trade: 62002 role: XMR seller as taker
Jul-15 23:11:59:36 (+0 ms)	[L0] Trade: 87102 role: XMR seller as taker
Jul-15 23:11:59:36 (+0 ms)	[L0] Trade: 74839807 role: XMR seller as taker

 PASS  src/HavenoClient.test.ts (58.061 s)
  ✓ Do all trades contain role property (302 ms)




Jul-15 23:21:32:60 (+3395 ms)	[L0] BEFORE TEST "Do all trades contain role property"
Jul-15 23:21:32.682 [grpc-default-executor-0] INFO  h.d.g.i.CallRateMeteringInterceptor: GetTrades has been called 1 time in the last second, rate limit is 10/second 
Jul-15 23:21:32.830 [grpc-default-executor-0] INFO  h.d.g.i.CallRateMeteringInterceptor: GetTrades has been called 2 times in the last second, rate limit is 10/second 
Jul-15 23:21:32.893 [grpc-default-executor-0] INFO  h.d.g.i.CallRateMeteringInterceptor: GetTrades has been called 1 time in the last second, rate limit is 10/second 
Jul-15 23:21:32:86 (+326 ms)	[L0] Trades size: 6
Jul-15 23:21:32:86 (+0 ms)	[L0] Trade: 62002 role: 
Jul-15 23:21:32:86 (+0 ms)	[L0] Trade: 87102 role: 
Jul-15 23:21:32:87 (+1 ms)	[L0] Trade: 74839807 role: 
Jul-15 23:21:32:87 (+0 ms)	[L0] Trade: 62002 role: 
Jul-15 23:21:32:87 (+0 ms)	[L0] Trade: 87102 role: 
Jul-15 23:21:32:87 (+0 ms)	[L0] Trade: 74839807 role: 

 FAIL  src/HavenoClient.test.ts (56.519 s)
  ✕ Do all trades contain role property (378 ms)
```

### CODE REVIEW

**Comment by woodser** [comment]
Instead of adding a new test for this specific scenario, better to call getTrades() after the offer is taken in `executeTrade`, test that the trade exists by ID, and test the role field for each trade in `testTrade()`

**Comment by fa2a5qj3** [comment]
A problem happens, because `testTrade()` is called from multiple places, and the TradeInfo object returned from `havenod.takeOffer()` contains an empty/null role field therefore failing the test early.  I'm not sure if that means takeOffer needs fixing, or the test needs adjusting.

File: src/HavenoClient.test.ts [hunk 1]
```diff
@@ -2816,6 +2821,7 @@ async function takeOffer(ctxP: Partial<TradeContext>): Promise<TradeInfo> {
   }
 
   // test trade model
+  HavenoUtils.log(0, "TradeInfo received from havenod.takeOffer: " + trade.getShortId() + " role: " + trade.getRole());
```
Comment (woodser) [inline src/HavenoClient.test.ts:2824]: No need for comment.

File: src/HavenoClient.test.ts [hunk 2]
```diff
@@ -2877,6 +2883,13 @@ async function testTrade(trade: TradeInfo, ctx: TradeContext) {
   expect(BigInt(trade.getBuyerSecurityDeposit())).toEqual(expectedSecurityDeposit - ctx.getBuyer().depositTxFee);
   expect(BigInt(trade.getSellerSecurityDeposit())).toEqual(expectedSecurityDeposit - ctx.getSeller().depositTxFee);
 
+  HavenoUtils.log(0, "Trade: " + trade.getShortId() + " role: " + trade.getRole());
```
Comment (woodser) [inline src/HavenoClient.test.ts:2886]: No need for comment.

File: src/HavenoClient.test.ts [hunk 3]
```diff
@@ -2600,6 +2600,11 @@ async function executeTrade(ctxP: Partial<TradeContext>): Promise<string> {
       ctx.taker.balancesAfterPayout = await ctx.taker.havenod?.getBalances();
     }
 
+    const trades = (await user1.getTrades()).concat(await user2.getTrades());
```
Comment (woodser) [inline src/HavenoClient.test.ts:2603]: Let's remove this here, and [here](https://github.com/haveno-dex/haveno-ts/blob/7192b899f493702547ce4ddce4dd1b5b3497f822/src/HavenoClient.test.ts#L2854) in takeOffer:

- Get user1 trades, assert the trade is contained by id, and call testTrade() on each.
- Get user2 trades, assert the trade is contained by id, and call testTrade() on each.

**Comment by woodser** [comment]
> A problem happens, because testTrade() is called from multiple places, and the TradeInfo object returned from havenod.takeOffer() contains an empty/null role field therefore failing the test early

Sorry I missed this. Sounds like the the returned `TradeInfo` from `takeOffer()` needs fixed to return this field. This issue is a good chance to ensure the field is always correct.

**Comment by woodser** [comment]
https://github.com/haveno-dex/haveno-ts/pull/301 opened as an alternative and I added you as a co-author.

This API addition and test refactoring was a little tricker than expected.

**Pull request closed without merging**